content/blog/2020-12-20-bril-algebraic-data-types/index.md

```diff
@@ -0,0 +1,538 @@
+
+The JSON format of the `construct` instruction is as follows:
+
+```json
+{
+    "op": "construct",
+    "args": ["a"],
+    "dest": "d",
+    "type": {"sum": [t1, ..., tn]}
+}
+```
+
+The text format of the `construct` instruction is as follows:
+
+```
+d: sum<t1, ..., tn> = construct a
+```
+
+In theory, it could have been possible to omit `construct` entirely if
+there were a sum type subtyping rule in place, but such a rule would
+probably have made things more, not less, confusing.
+
+#### `destruct`
+
+The `destruct` instruction is used to execute different code based on
+the constructor used to create a sum-typed value. It takes a sum-typed
+value as an argument and a number of labels, one for each constructor
+in the sum type.
+
+The destination of the instruction is a variable name with a sum
+type. The semantics of the instruction dictate that the control flow
+jumps to the correct branch based on the argument and that the
+destination value gets assigned the value within the constructor.
+
+The destination will in fact not have the same sum type as the
+argument; it will have the type carried by one of the constructors,
+which is different for each label. This is confusing because the
+destination is annotated with a type that it cannot have, but I chose
+this approach because the Bril interpreter needs to know the available
+constructors in order to select the correct label to jump to.
+
+The JSON format of the `destruct` instruction is as follows:
+
+```json
+{
+    "op": "destruct",
+    "args": ["a"],
+    "labels": ["l1", ..., "ln"],
+    "dest": "d",
+    "type": {"product": [t1, ..., tn]}
+}
+```
+
+The text format of the `destruct` instruction is as follows:
+
+```
+d: sum<t1, ..., tn> = destruct v .l1 ... .ln
+```
+
+A common use case is some way of representing an empty, or unit,
+value. For this it is recommended to use the empty product type,
+i.e. `product<>`. For example, to represent an optional `int`, one
+could use the following type: `sum<int, product<>>`.
+
+I debated whether to even include a powerful instruction like
+`destruct` at all. I could have instead offered an instruction to get
+the index of the constructor used for a value and another instruction
+for casting a sum type down to one of the constructor-carried
+types, using a series of branch instructions to replicate the behavior
+of `destruct`.
+
+The reason that I opted for the `destruct` approach is that it makes
+it possible, and not terribly difficult, to type-check the usage of
+sum types in a Bril program. With the alternative approach, it is not
+clear to me that it is even possible to perform type-checking of
+things like the casting instruction described above. `destruct` does
+seem rather large and clunky for an IR, especially compared to the
+other instructions (it's like a whole bunch of `br` instructions in
+one!), but I think the ability to perform type-checking makes it
+worthwhile.
```

Comment:
Well, you could type-check it, but you couldn't guarantee that the program did bad projections. Which is maybe appropriate for an IR but not for a higher-level language? Interesting design question.